### PR TITLE
Add GPU kernel for SparseApplyFtrl

### PR DESCRIPTION
**Pull request by benbarsdell**
Also applies to Resource and V2 versions of the op.

cc @sanjoy @nluehr

### CODE REVIEW

File: tensorflow/core/kernels/training_ops.cc [hunk 1]
```diff
@@ -2761,7 +2761,8 @@ class SparseApplyFtrlOp : public OpKernel {
     OP_REQUIRES(
         ctx,
         TensorShapeUtils::IsScalar(lr.shape()) &&
-            (lr.scalar<T>()() > static_cast<T>(0) ||
+            (!std::is_same<Device, CPUDevice>::value ||
+             lr.scalar<T>()() > static_cast<T>(0) ||
```
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.cc:2765]: On GPU what happens if these constraints are violated?  Do we produce a bogus answer?  Or do crash in the CUDA kernel?

(Please also add the answer as a comment.)

**Review by sanjoy — CHANGES_REQUESTED** [review]
Have you verified that the new kernels you've added are tested?  I.e. if you add a `LOG(FATAL)` in the GPU kernel are there tests that fail?

File: tensorflow/core/kernels/training_ops.cc (hunk 1 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops.cc:2769]: The kernel won't crash, just give bogus results. Comment added.

**Comment by benbarsdell** [comment]
I confirmed that the GPU kernel is being called in the CI tests.

File: tensorflow/core/kernels/training_ops.cc (hunk 1 above)
Comment (Naman528) [reply tensorflow/core/kernels/training_ops.cc:2769]: The kernel won't crash, just give bogus results.

File: tensorflow/core/kernels/training_ops_gpu.cu.cc [hunk 2]
```diff
@@ -573,6 +639,32 @@ struct ApplyFtrlV2MultiplyLinearByLr<GPUDevice, T> {
...
+    const Tindex first_dim_size = var.dimension(0);
+    const Tindex grad_size = grad.size();
+    const Tindex indices_size = indices.size();
+    GpuLaunchConfig config = GetGpuLaunchConfig(grad_size, d);
+    return GpuLaunchKernel(
+        SparseApplyFtrlKernel<T, Tindex, has_l2_shrinkage>, config.block_count,
+        config.thread_per_block, 0, d.stream(), var.data(), accum.data(),
+        linear.data(), lr.data(), l1.data(), l2.data(), l2_shrinkage.data(),
+        lr_power.data(), grad.data(), indices.data(), first_dim_size, grad_size,
+        indices_size, multiply_linear_by_lr);
```
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops_gpu.cu.cc:664]: Can you please add comments of the form `/*l1=*/` at the right places.  Internally we have a clang check that ensures that the comment match the parameter names which avoids bugs from e.g. swapping `l1.data()` and `l2.data()`.

**Review by sanjoy — CHANGES_REQUESTED** [review]
Sorry for letting this slip through the cracks.

The code LGTM, only one minor comment.

Also the comment on the class says `// Note, this op works on cpu only.` which should be fixed.

File: tensorflow/core/kernels/training_ops_gpu.cu.cc (hunk 2 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops_gpu.cu.cc]: Done (also removed the "cpu only" comment).
(Will the clang check work through the GpuLaunchKernel wrapper function?)

**Review by sanjoy — APPROVED** [review]

**Pull request merged**